File: lib/base/setmetatable.go
package base

import (
	rt "github.com/arnodel/golua/runtime"
)

func setmetatable(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.CheckNArgs(2); err != nil {
		return nil, err
	}
	tbl, err := c.TableArg(0)
	if err != nil {
		return nil, err
	}
	if !rt.RawGet(tbl.Metatable(), rt.StringValue("__metatable")).IsNil() {
		return nil, rt.NewErrorS("cannot set metatable")
	}
	if c.Arg(1).IsNil() {
		tbl.SetMetatable(nil)
	} else if meta, err := c.TableArg(1); err == nil {
		tbl.SetMetatable(meta)
	} else {
		return nil, err
	}
	return c.PushingNext1(t.Runtime, c.Arg(0)), nil
}

File: lib/base/lua/setmetatable.lua
local function test(...)
    local ok, val = pcall(setmetatable, ...)
    if ok then
        print("ok")
    else
        print(val)
    end
end

test()
--> ~.*: 2 arguments needed

test(1)
--> ~.*: 2 arguments needed

test(1, 2)
--> ~.*must be a table

t = {}
setmetatable(t, {__metatable={}})
test(t, nil)
--> ~.*: cannot set metatable

t = {}
test(t, nil)
--> =ok

test(t, 123)
--> ~.*must be a table
